src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -153,24 +167,126 @@ void GcsPlacementGroupManager::HandleCreatePlacementGroup(
+      [this, request, reply, send_reply_callback, placement_group_id,
+       placement_group](Status status) {
         RAY_CHECK_OK(status);
-        RegisterPlacementGroup(request, [reply, send_reply_callback, placement_group_id,
-                                         name, strategy]() {
-          RAY_LOG(INFO) << "Finished registering placement group, placement group id = "
-                        << placement_group_id << ", name = " << name
-                        << ", strategy = " << strategy;
-          GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+        if (registered_placement_groups_.find(placement_group_id) ==
+            registered_placement_groups_.end()) {
+          std::stringstream stream;
+          stream << "Placement group of id " << placement_group_id
+                 << " has been removed before registration.";
+          GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::NotFound(stream.str()));
+          return;
+        }
+
+        RegisterPlacementGroup(
+            request, [reply, send_reply_callback, placement_group](Status status) {
+              if (status.ok()) {
+                RAY_LOG(INFO) << "Finished registering placement group, "
+                              << placement_group->DebugString();
+              } else {
+                RAY_LOG(INFO) << "Failed to register placement group, "
+                              << placement_group->DebugString();
+              }
+              GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+            });
+      }));
+}
+
+void GcsPlacementGroupManager::HandleRemovePlacementGroup(
+    const rpc::RemovePlacementGroupRequest &request,
+    rpc::RemovePlacementGroupReply *reply, rpc::SendReplyCallback send_reply_callback) {
+  const auto placement_group_id =
+      PlacementGroupID::FromBinary(request.placement_group_id());
+
+  RemovePlacementGroup(placement_group_id, [send_reply_callback, reply,
+                                            placement_group_id](Status status) {
+    if (status.ok()) {
+      RAY_LOG(INFO) << "Placement group of an id, " << placement_group_id
+                    << " is removed successfully.";
+    } else {
+      RAY_LOG(INFO) << "Removing a placement group of an id, " << placement_group_id
+                    << " failed.";
+    }
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  });
+}
+
+void GcsPlacementGroupManager::RemovePlacementGroup(
+    const PlacementGroupID &placement_group_id,
+    StatusCallback on_placement_group_removed) {
+  RAY_CHECK(on_placement_group_removed);
+  // If the placement group is not registered, it means it is already removed or wasn't
+  // created before. Just fail the request in this case.
+  if (registered_placement_groups_.find(placement_group_id) ==
+      registered_placement_groups_.end()) {
+    RAY_CHECK(registered_placement_groups_.find(placement_group_id) ==
+              registered_placement_groups_.end());
```

Comment:
Why check twice?

Reply:
Yeah that's a good catch. Idk what's the best solution in this case. The issue is we need to emplace it before we call Put to avoid timing issues. But, we need it inside this method because unit tests need them.

Reply:
Removed a second check.